src/pexprs-madlibs.js

```diff
@@ -0,0 +1,209 @@
+
+pexprs.Seq.prototype.generateExample = function(grammar, examples, inSyntacticContext, actuals) {
+  var factorExamples = this.factors.map(function(factor) {
+    return factor.generateExample(grammar, examples, inSyntacticContext, actuals);
+  });
+  var categorizedExamples = categorizeExamples(factorExamples);
+
+  var examplesNeeded = categorizedExamples.examplesNeeded;
+  var successfulExamples = categorizedExamples.successfulExamples;
+  var needHelp = categorizedExamples.needHelp;
+
+  var returnObj = {};
+
+  // in a Seq, all pieces must succeed in order to have a successful example
+  if (examplesNeeded.length > 0 || needHelp) {
+    returnObj.examplesNeeded = examplesNeeded;
+    returnObj.needHelp = needHelp;
+  } else {
+    returnObj.example = successfulExamples.join(inSyntacticContext ? ' ' : '');
+  }
+
+  return returnObj;
+};
+
+pexprs.Apply.prototype.generateExample = function(grammar, examples, inSyntacticContext, actuals) {
+  var returnObj = {};
+
+  var ruleName = this.substituteParams(actuals).toString();
+
+  if (!examples.hasOwnProperty(ruleName)) {
+    returnObj.examplesNeeded = [ruleName];
+  } else {
+    var relevantExamples = examples[ruleName];
+    var i = Math.floor(Math.random() * relevantExamples.length);
+    returnObj.example = relevantExamples[i];
+  }
+
+  return returnObj;
+};
+
+// assumes that terminal's object is always a string
+pexprs.Terminal.prototype.generateExample = function(grammar, examples, inSyntacticContext) {
+  return {example: this.obj};
+};
+
+pexprs.Range.prototype.generateExample = function(grammar, examples, inSyntacticContext) {
+  var rangeSize = this.to.charCodeAt(0) - this.from.charCodeAt(0);
+  return {example: String.fromCharCode(
+    this.from.charCodeAt(0) + Math.floor(rangeSize * Math.random())
+  )};
+};
+
+// these all fail for now
+pexprs.Not.prototype.generateExample = function(grammar, examples, inSyntacticContext) {
+  return {example: ''};
+};
+
+pexprs.Lookahead.prototype.generateExample = function(grammar, examples, inSyntacticContext) {
+  return {example: ''};
+};
+
+pexprs.Param.prototype.generateExample = function(grammar, examples, inSyntacticContext, actuals) {
+  return actuals[this.index].generateExample(grammar, examples, inSyntacticContext, actuals);
+};
+
+function repeat(n, fn) {
+  if (n < 0) { return; }
+  while (n > 0) {
+    fn(n);
+    n--;
+  }
+}
+
+var generateNExamples = function(that, grammar, examples, inSyntacticContext, actuals, numTimes) {
+  var items = [];
+  repeat(numTimes, function() {
+    items.push(that.expr.generateExample(grammar, examples, inSyntacticContext, actuals));
+  });
+
+  var categorizedExamples = categorizeExamples(items);
+
+  var examplesNeeded = categorizedExamples.examplesNeeded;
+  var successfulExamples = categorizedExamples.successfulExamples;
+
+  var returnObj = {};
+
+  // it's always either one or the other
+  returnObj.example = successfulExamples.join(inSyntacticContext ? ' ' : '');
```

Comment:
Hmmm, this should probably be calling `spaces.generateExample` here, since it's possible to override `space` such that `' '` is not a valid example! It's fine to leave that for a follow-up commit though, so maybe just put a TODO in here.